### ISSUE TITLE
Timer benchmark stores only one time value, and therefore has broken mean/median/etc metrics

### ISSUE THREAD

**Issue by HugoTessier** [pytorch/pytorch#106801]
### 🐛 Describe the bug

Here is a simple code to reproduce the error:

```
import torch
import torchvision
import torch.utils.benchmark as benchmark

if __name__ == '__main__':
    net = torchvision.models.resnet18().cuda()
    data = torch.rand((1, 3, 224, 224), device='cuda')

    net.eval()
    with torch.inference_mode():
        timer = benchmark.Timer(stmt='net(data)', num_threads=1, globals={'net': net, 'data': data})
        measurement = timer.timeit(number=10)
    print(measurement.raw_times, measurement.median, measurement.mean, measurement.iqr)
```

This should display a list of 10 values, different median and mean and non-zero iqr.
However, I get this:

```
[0.05537769995862618] 0.005537769995862618 0.005537769995862618 0.0
```

This means that only one value is stored, which not only prevents me doing what I wanted (i.e. get all the raw values and plot them on a graph), but also explains the broken behavior of the metrics (same mean and median, null iqr...).

I looked up the source code a bit, in torch/utils/benchmark/utils/timer.py, and found that:

```
    def _timeit(self, number: int) -> float:
        # Even calling a timer in C++ takes ~50 ns, so no real operation should
        # take less than 1 ns. (And this prevents divide by zero errors.)
        return max(self._timer.timeit(number), 1e-9)

    def timeit(self, number: int = 1000000) -> common.Measurement:
        """Mirrors the semantics of timeit.Timer.timeit().

        Execute the main statement (`stmt`) `number` times.
        https://docs.python.org/3/library/timeit.html#timeit.Timer.timeit
        """
        with common.set_torch_threads(self._task_spec.num_threads):
            # Warmup
            self._timeit(number=max(int(number // 100), 2))

            return common.Measurement(
                number_per_run=number,
                raw_times=[self._timeit(number=number)],
                task_spec=self._task_spec
            )
```

And, if I understand well, it means that, by definition, only one value will ever be stored.
I don't think this is the intended behavior at all, and therefore suggest either to fix the source code, or to update the documentation to make it clearer how it actually behaves.

### Versions

Collecting environment information...
PyTorch version: 2.1.0.dev20230726+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 10 Famille
GCC version: Could not collect
Clang version: Could not collect
CMake version: Could not collect
Libc version: N/A

Python version: 3.11.4 (tags/v3.11.4:d2340ef, Jun  7 2023, 05:45:37) [MSC v.1934 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-10-10.0.19045-SP0
Is CUDA available: True
CUDA runtime version: 12.2.91
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce GTX 1050
Nvidia driver version: 536.25
cuDNN version: C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v12.2\bin\cudnn_ops_train64_8.dll
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:

Revision=

Versions of relevant libraries:
[pip3] numpy==1.25.1
[pip3] torch==2.1.0.dev20230726+cu121
[pip3] torchaudio==2.1.0.dev20230727+cu121
[pip3] torchinfo==1.8.0
[pip3] torchvision==0.16.0.dev20230727+cu121
[conda] Could not collect

**Comment by drisspg** [comment]
While slightly confusing I do think that this does indeed match timer.timeit as specified in the documentation. I think what you want is something like repeat?  Which unfortunately is not supported by the timer class right now. However it should be pretty simple to create the same functionality by rapping the timer.timeit function in a for loop.

**Comment by HugoTessier** [comment]
You mean that the number of repetitions is not there to actually store n values and be able to do statistics on them?
And therefore, if there are repetitions, it is only to somehow keep the maximum value or something like that?
This is quite aweird choice, especially because the Measurement object offers to compute medians, mean, etc, which supposes multiple values, while by design, there can only be one...
Maybe there is a good reason behind this, but I remain skeptical.
Thanks for the answer, though.

**Comment by drisspg** [comment]
TBH This is kind of surprising to me as well.  When reading the documentation: https://pytorch.org/docs/stable/benchmark_utils.html it says `The PyTorch Timer is based on timeit.Timer (and in fact uses timeit.Timer internally), but with several key`

The key difference I can see is `To that effect, this class deviates from the timeit API by conceptually merging timeit.Timer.repeat and timeit.Timer.autorange.`

https://pytorch.org/docs/stable/benchmark_utils.html#torch.utils.benchmark.Timer.timeit explicitly mentions that the number argument is just for specifying the number of times the statement you are trying to measure is run, and the averaged time of all this number amounts to 1 experiment. You want to get statistics on N number of experiments, and the method for doing this `repeat` is disabled in the Pytorch benchmark utils. However if you did something like 

```
with torch.inference_mode():
        timer = benchmark.Timer(stmt='net(data)', num_threads=1, globals={'net': net, 'data': data})
        measurement = timer.blocked_autorange(min_run_time=10)
 ```
 I think you would see something closer to what you want, only you don't have direct control over the # of experiments ran.

**Comment by HugoTessier** [comment]
Yes, using blocked_autorange is what I just did, and it actually stores the value of every run, which is so much more convenient (even though, what is **not** convenient is that I therefore have a varying number of points per experiment).
So maybe the behavior of timeit is faithful to the intended design, but then it is quite a weird design choice.
IMHO, it would work as well, while being more convenient, for timeit to store all values, and be able to give either the total time, the mean, the minimum, etc, instead of storing a single arbitrary value that we cannot analyse finely...

**Comment by sixt99** [comment]
Hello, I agree that this is indeed a weird feature and should therefore be fixed. To me, timeit(10) should mean the experiment is run 10 times and time.mean should yield the mean of times of all those runs.
